src/app/codegen-data-model/CodegenDataModel_Write.cpp

```diff
@@ -321,45 +332,68 @@ CHIP_ERROR CodegenDataModel::WriteAttribute(const InteractionModel::WriteAttribu
+            ChipLogError(DataManagement, "Unable to get cluster info for Endpoint 0x%x, Cluster " ChipLogFormatMEI,
                          request.path.mEndpointId, ChipLogValueMEI(request.path.mClusterId));
-            return CHIP_ERROR_NOT_FOUND;
+            return CHIP_IM_GLOBAL_STATUS(DataVersionMismatch);
         }
 
         if (request.path.mDataVersion.Value() != clusterInfo->dataVersion)
         {
-            ChipLogError(DataManagement, "Write Version mismatch for Endpoint %x, Cluster " ChipLogFormatMEI,
+            ChipLogError(DataManagement, "Write Version mismatch for Endpoint 0x%x, Cluster " ChipLogFormatMEI,
                          request.path.mEndpointId, ChipLogValueMEI(request.path.mClusterId));
             return CHIP_IM_GLOBAL_STATUS(DataVersionMismatch);
         }
     }
 
     AttributeAccessInterface * aai       = GetAttributeAccessOverride(request.path.mEndpointId, request.path.mClusterId);
     std::optional<CHIP_ERROR> aai_result = TryWriteViaAccessInterface(request.path, aai, decoder);
-
     if (aai_result.has_value())
     {
+        if (*aai_result == CHIP_NO_ERROR)
+        {
+            // TODO: change callbacks should likely be routed through the context `MarkDirty` only
+            //       however for now this is called directly because ember code does this call
+            //       inside emberAfWriteAttribute.
+            MatterReportingAttributeChangeCallback(request.path);
+            CurrentContext().dataModelChangeListener->MarkDirty(request.path);
+        }
         return *aai_result;
     }
 
-    ReturnErrorCodeIf(aai_result.has_value(), *aai_result);
+    MutableByteSpan dataBuffer = gEmberAttributeIOBufferSpan;
+    ReturnErrorOnFailure(DecodeValueIntoEmberBuffer(decoder, *attributeMetadata, dataBuffer));
 
-    ReturnErrorOnFailure(DecodeValueIntoEmberBuffer(decoder, *attributeMetadata, gEmberAttributeIOBufferSpan));
+    Protocols::InteractionModel::Status status;
 
-    EmberAfAttributeSearchRecord record;
-    record.endpoint    = request.path.mEndpointId;
-    record.clusterId   = request.path.mClusterId;
-    record.attributeId = request.path.mAttributeId;
+    if (request.operationFlags.Has(InteractionModel::OperationFlags::kInternal))
+    {
+        // Internal requests use the non-External interface that has less enforcement
+        // than the external version (e.g. does not check/enforce writable settings, does not
+        // validate atribute types) - see attribute-table.h documentation for details.
+        status = emberAfWriteAttribute(request.path.mEndpointId, request.path.mClusterId, request.path.mAttributeId,
+                                       gEmberAttributeIOBufferSpan.data(), (*attributeMetadata)->attributeType);
+    }
+    else
+    {
+        if (dataBuffer.size() > (*attributeMetadata)->size)
+        {
+            ChipLogDetail(Zcl, "Data to write exceedes the attribute size claimed.");
```

Comment:
nit: "exceeds".

Also, why is it OK to do this check on non-internal, but expect internal clients to do the right thing? When is it that `dataBuffer.size() > (*attributeMetadata)->size)` ? What about if `<` instead of `==`. The underlying conversion assumptions seem swallowed by the evil of `emAfWriteAttributeExternal` complex internal encoding rules. In other words, by looking at this code, it's really not clear why we would ever care to pre-check the dataBuffer.size() rather than let `emAfWriteAttributeExternal` tell us of errors if its assumptions are broken.

Reply:
This is copied over from https://github.com/project-chip/connectedhomeip/blob/master/src/app/util/ember-compatibility-functions.cpp#L798 (including the typo). I fixed the typo here, however I would prefer not to try to expand the scope of the PR with altering other existing logic. It is very large as is.

The ">" would be when you receive a string that is larger than what ember allocated in it's attribute store. so Say a char_string<4> would have 5 bytes allocated for ember (length + 4 chars, no null terminator). As a result, writhing "a" or "ab" is ok and the size is shorter. writing "abcd" fits exactly (the == case) and "abcdef" will result in the error here.

I guess TLDR: I am maintaining functionality of ember-compatibility-functions.cpp here, so all stays reasonably similar unless I have strong reasons to do it differently. I cannot actually reference "ember-compatibiltiy-functions.cpp" in comments because I expect that file will eventually be deleted if we integrate this code (as it does the same thing)